src/wrappers/themis/python/pythemis/scell.py

```diff
@@ -70,11 +101,99 @@ def decrypt(self, message, context=None):
+        elif isinstance(passphrase, type(b'')):
+            pass
+        else:
+            raise SecureCellError('passphrase must be either "unicode" or "bytes"')
+        self.passphrase = passphrase
+
+    def encrypt(self, message, context=None):
+        """
+        Encrypt given message with optional context.
+
+        :type message: bytes
+        :type context: bytes
+        :returns bytes
+        """
+        context_length = len(context) if context else 0
+        encrypted_message_length = c_int(0)
+
+        res = themis.themis_secure_cell_encrypt_seal_with_passphrase(
+                self.passphrase, len(self.passphrase),
+                context, context_length,
+                message, len(message),
+                None, byref(encrypted_message_length))
+        if res != THEMIS_CODES.BUFFER_TOO_SMALL:
+            raise SecureCellError("encryption failed", error_code=res)
+
+        encrypted_message = create_string_buffer(encrypted_message_length.value)
+        res = themis.themis_secure_cell_encrypt_seal_with_passphrase(
+                self.passphrase, len(self.passphrase),
+                context, context_length,
+                message, len(message),
+                encrypted_message, byref(encrypted_message_length))
+        if res != THEMIS_CODES.SUCCESS:
+            raise SecureCellError("encryption failed", error_code=res)
+
+        return string_at(encrypted_message, encrypted_message_length.value)
+
+    def decrypt(self, message, context=None):
+        """
+        Decrypt given message with optional context.
+
+        :type message: bytes
+        :type context: bytes
+        :returns bytes
+        """
+        context_length = len(context) if context else 0
+        decrypted_message_length = c_int(0)
+
+        res = themis.themis_secure_cell_decrypt_seal_with_passphrase(
+                self.passphrase, len(self.passphrase),
+                context, context_length,
+                message, len(message),
+                None, byref(decrypted_message_length))
+        if res != THEMIS_CODES.BUFFER_TOO_SMALL:
+            raise SecureCellError("decryption failed", error_code=res)
+
+        decrypted_message = create_string_buffer(decrypted_message_length.value)
+        res = themis.themis_secure_cell_decrypt_seal_with_passphrase(
+                self.passphrase, len(self.passphrase),
+                context, context_length,
+                message, len(message),
+                decrypted_message, byref(decrypted_message_length))
+        if res != THEMIS_CODES.SUCCESS:
+            raise SecureCellError("decryption failed", error_code=res)
+
+        return string_at(decrypted_message, decrypted_message_length.value)
+
+
 class SCellTokenProtect(object):
     def __init__(self, key):
         if not key:
             raise ThemisError(THEMIS_CODES.FAIL,
                               "Secure Cell (Token Protect) failed creating")
+        if not isinstance(key, type(b'')):
```

Comment:
> why not just `isinstance(key, bytes)`?

Because of Python 2 compatibility.

Yes, Python 2.7 defines `bytes` as a synonym for `str`, but unfortunately Python 3 does _not_ define `unicode` as a synonym for `str`, and we need to check for Unicode strings when working with passphrases.

My previous approach was to add something like

```python
__pythemis_bytes__ = str if sys.version_info[0] == 2 else bytes
__pythemis_str__ = unicode if sys.version_info[0] == 2 else str
```

to avoid computing types every time, but I figured it to be more ugly.

Another thing that came to my mind was to use `isinstance(key, bytes)` to check for keys (which will work in both Python 2 and 3) and for passphrases do something like

```python
if not isinstance(passphrase, bytes):
    # If not bytes, assume that it's a string.
    # If it's not a string either, you'll get an AttributeError.
    passphrase = passphrase.encode(encoding)
```

but I thought that having better error messages about type mismatch is more important.

What would you advise?

Reply:
one more popular way for libraries that want to support py2/3 is to use https://six.readthedocs.io/ library which provide such set of types/checks, like `six.binary_type/six.text_type`. It's okay to add six to dependencies in `setup.py` and use it.
p.s. popular web framework [django](https://docs.djangoproject.com/en/1.11/topics/python3/#writing-compatible-code-with-six)  used custom version of six for py2/3 compatibility with own helpers

Reply:
> It's okay to add `six` to dependencies in setup.py and use it.

Oh... For some reason I thought we'd better not add dependencies. I'll look into it. Thanks.

Reply:
it's good to not have dependencies if we can. but as I remember, we had some dependency for enum types and their compatibility in py2/3. plus if it's small piece of code, we can just copy it. anyway this code will not changed in the future because python2 is going out from usage in more projects each year. probably after 1-2 years we just remove all of these hacks and py2 support.